proto/kuksa/val/v2/val.proto

```diff
@@ -0,0 +1,146 @@
+/********************************************************************************
+ * Copyright (c) 2022 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Apache License 2.0 which is available at
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ********************************************************************************/
+
+syntax = "proto3";
+
+package kuksa.val.v2;
+
+option go_package = "kuksa/val/v2";
+
+import "types.proto";
+
+// Note on authorization:
+// Tokens (auth-token or auth-uuid) are sent as (GRPC / http2) metadata.
+//
+// The auth-token is a JWT compliant token as the examples found here:
+// https://github.com/eclipse-kuksa/kuksa-databroker/tree/main/certificates/jwt
+//
+// See also https://github.com/eclipse-kuksa/kuksa-databroker/blob/main/doc/authorization.md#jwt-access-token
+//
+// Upon reception of auth-token, server shall generate an auth-uuid in metadata
+// that the client can use instead of auth-token in subsequent calls.
+
+service VAL {
+  // get current value of a single attribute, sensor or actuator
+  rpc Get(GetRequest) returns (GetResponse);
+  
+  // get current values of a set of attributes, sensors or actuators
+  rpc GetBatch(GetBatchRequest) returns (GetBatchResponse);
+
+  // subscribe to sensor and actuator values
+  rpc Subscribe(SubscribeRequest) returns (stream SubscribeResponse);
+
+  // actuate on a single actuator value
+  rpc Actuate(ActuateRequest) returns (ActuateResponse);
+
+  // actuate on more than one actuators values
+  rpc ActuateBatch(ActuateBatchRequest) returns (ActuateBatchResponse);
+  
+  // provide current value of attributes, sensors and actuator - for low update frequency
+  rpc ProvideOnce(ProvideOnceRequest) returns (ProvideOnceResponse);
+
+  // provide stream of current values of attributes, sensors and actuator - for high frequency
+  // return stream of ACK
+  rpc ProvideStream(stream ProvideUpwardsStream) returns (stream ProvideDownwardsStream);
+
```

Comment:
I still have some troubles with the name of this function. "ProvideStream" would mean that "a stream is provided", but that does match the sense of it.
What about: `StreamedProvide`/`StreamProvide`?
Maybe also analogous to the name pairing suggested by John (`Get` and `BatchGet`) we could have `Provide` and `Stream(ed)Provide` ...

Regarding the request/response message naming:
Basically, I agree with John. On the other hand, as we have to streams here the "classic" pattern of "one" request and "one" response isn't matching here. Especially, as both streams (to and from broker) can transport requests and responses (one the layer on top of gRPC).
So, I think it would be worth to reflect that in the naming for this special case, e.g. `ProviderStreamToBroker` and `...FromBroker`.

Reply:
> I still have some troubles with the name of this function. "ProvideStream" would mean that "a stream is provided", but that does match the sense of it.

Do you also have a problem with `ProvideOnce`?

As I see it, this is just a way to differentiate between the [unary vs streaming](https://grpc.io/docs/what-is-grpc/core-concepts/#unary-rpc) version of the call, i.e. `ProvideUnary` and `ProvideStreaming`. And `ProvideOnce` is perhaps easier to understand than `ProvideUnary`.

If we want, we can use `Provide` for the streaming version (as in the "original proposal").

> What about: StreamedProvide/StreamProvide?

I think these just sounds weird, and are not the logical counterpart to `ProvideOnce` or `ProvideUnary`. Then it's better to just use `Provide` or `ProvideStreaming` in my opinion (or `ProvideStream` for that matter).

Reply:
> As I see it, this is just a way to differentiate between the [unary vs streaming](https://grpc.io/docs/what-is-grpc/core-concepts/#unary-rpc) version of the call, i.e. ProvideUnary and ProvideStreaming.

Compare if you want with the practice of adding a suffix to differentiate between async and sync version of functions in APIs.

I.e. `GetFoo` vs `GetFooAsync`.

Reply:
> Do you also have a problem with ProvideOnce?
>
> As I see it, this is just a way to differentiate between the [unary vs streaming](https://grpc.io/docs/what-is-grpc/core-concepts/#unary-rpc) version of the call, [...]

Yes, that's not my point. I've just issues with the name `ProvideStream`. It is not about "providing a stream" but about doing a continuous providing. Also grammatically, "ProvideStream" sounds/reads strange as "once" is an adverb whereas "stream" is a noun. So, the counterpart for "provide (something) once" would be "provide (something) streamed".

But if that sound weird, what about `ProvideContinuously`, `...Periodicly`,  `...Repeated(ly)`, `...Frequently`?

Reply:
Yes, I would agree that `Stream` is not the counterpart to `Unary` or `Once`.

But I don't thing `Streamed` is the counterpart either as that is in the past tense (which is one reason I think it sounds weird).

The counterpart to `Unary` is `Streaming` - not `Stream` or `Streamed` [(unary vs streaming)](https://grpc.io/docs/what-is-grpc/core-concepts/#unary-rpc).

Or we can just call it `Provide` and we don't have to figure out which other word to add :smile:

Reply:
Yes, this function is really "special" - an "Eier-legende-Wollmilchsau" (egg-dumping wool-milk-sow) ;-)
(And yes, "streamed" is wrong - my mistake. Had some wrong grammar in my head.)

Basically it's about "establishing a stream between provider and broker" - so: `OpenProviderStream`, maybe?

Reply:
> Basically it's about "establishing a stream between provider and broker" - so: OpenProviderStream, maybe?

That doesn't sound terrible. But I don't think `Provide` sounds terrible either (and it's shorter). I guess I prefer both `OpenProviderStream` and `Provide` over the alternatives so far.